### PR TITLE
Fix maximum page size estimate in Parquet writer

### PR DESCRIPTION
**Pull request by vuule**
## Description
Closes https://github.com/rapidsai/cudf/issues/11916

cuda memcheck reports an OOB write in one of the tests. The root cause is an underallocated buffer for encoded pages.
This PR fixes the computation of the maximum size of data pages (RLE encoded) when dictionary encoding is used.
Other changes:
Refactored max RLE page size computation to avoid code repetition.
Use actual dictionary index width instead of (outdated) worst case.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/page_enc.cu [hunk 1]
```diff
@@ -343,8 +351,8 @@ __global__ void __launch_bounds__(128)
           (values_in_page > 0 && (page_size + fragment_data_size > this_max_page_size)) ||
           rows_in_page >= max_page_size_rows) {
         if (ck_g.use_dictionary) {
-          page_size =
-            1 + 5 + ((values_in_page * ck_g.dict_rle_bits + 7) >> 3) + (values_in_page >> 8);
+          // Additional byte to store entry bit width
+          page_size = 1 + max_RLE_page_size(ck_g.dict_rle_bits, values_in_page);
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:355]: Based on [Apache Parquet format docs](https://parquet.apache.org/docs/file-format/data-pages/encodings/#dictionary-encoding-plain_dictionary--2-and-rle_dictionary--8):
> Data page format: the bit width used to encode the entry ids stored as 1 byte (max bit width = 32), followed by the values encoded using RLE/Bit packed described above (with the given bit width).

File: cpp/src/io/parquet/page_enc.cu [hunk 2]
```diff
@@ -217,6 +217,14 @@ __global__ void __launch_bounds__(128)
   if (frag_id < num_fragments_per_column and lane_id == 0) groups[column_id][frag_id] = *g;
 }
 
+constexpr uint32_t max_RLE_page_size(uint8_t value_bit_width, uint32_t num_values)
```
Comment (bdice) [inline cpp/src/io/parquet/page_enc.cu:220]: Nice refactoring here.

File: cpp/src/io/parquet/page_enc.cu (hunk 1 above)
Comment (bdice) [reply cpp/src/io/parquet/page_enc.cu:355]: This is the bug fix and the rest of the diff is refactoring, right? Is a C++ test or Python comparison to another reader/writer needed?

Comment (davidwendt) [reply cpp/src/io/parquet/page_enc.cu:355]: Technically there is already a gtest for this. The error was an OOB write and only manifested with a memcheck without an rmm pool which is tested nightly.

**Review by davidwendt — APPROVED** [review]

File: cpp/src/io/parquet/page_enc.cu (hunk 1 above)
Comment (bdice) [reply cpp/src/io/parquet/page_enc.cu:355]: Good call. Approving.

**Review by bdice — APPROVED** [review]

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**